src/main/java/com/iota/iri/network/TransactionRequester.java

```diff
@@ -103,61 +80,32 @@ void popEldestTransactionToRequest() {
 
     /**
      * This method allows to check if a transaction was requested by the TransactionRequester.
-     *
+     * <p>
      * It can for example be used to determine if a transaction that was received by the node was actively requested
      * while i.e. solidifying transactions or if a transaction arrived due to the gossip protocol.
      *
-     * @param transactionHash hash of the transaction to check
-     * @param milestoneRequest flag that indicates if the hash was requested by a milestone request
+     * @param transactionHash  hash of the transaction to check
      * @return true if the transaction is in the set of transactions to be requested and false otherwise
      */
-    public boolean isTransactionRequested(Hash transactionHash, boolean milestoneRequest) {
-        return (milestoneRequest && milestoneTransactionsToRequest.contains(transactionHash))
-                || (!milestoneRequest && milestoneTransactionsToRequest.contains(transactionHash) ||
-                transactionsToRequest.contains(transactionHash));
+    public boolean isTransactionRequested(Hash transactionHash) {
+        return transactionsToRequest.contains(transactionHash);
     }
 
     private boolean transactionsToRequestIsFull() {
         return transactionsToRequest.size() >= TransactionRequester.MAX_TX_REQ_QUEUE_SIZE;
     }
 
 
-    public Hash transactionToRequest(boolean milestone) throws Exception {
-        // determine which set of transactions to operate on
-        Set<Hash> primarySet = milestone ? milestoneTransactionsToRequest : transactionsToRequest;
-        Set<Hash> alternativeSet = milestone ? transactionsToRequest : milestoneTransactionsToRequest;
-        Set<Hash> requestSet = primarySet.size() == 0 ? alternativeSet : primarySet;
-
+    public Hash transactionToRequest() throws Exception {
         // determine the first hash in our set that needs to be processed
         Hash hash = null;
         synchronized (syncObj) {
             // repeat while we have transactions that shall be requested
-            while (requestSet.size() != 0) {
+            if (transactionsToRequest.size() != 0) {
                 // get the first item in our set for further examination
-                Iterator<Hash> iterator = requestSet.iterator();
+                Iterator<Hash> iterator = transactionsToRequest.iterator();
                 hash = iterator.next();
-
-                // if we have received the transaction in the mean time ....
-                if (TransactionViewModel.exists(tangle, hash)) {
```

Comment:
So you're saying that our chances of receiving a tx that we already have are so low, that it will be a shame to incur I/O costs?

I definitely agree, especially now that we only request what we are missing

Reply:
Yes, no need to do any I/O.